internal/provisioner/objects/dataplane/dataplane.go

```diff
@@ -431,8 +431,16 @@ func desiredDeployment(contour *model.Contour, contourImage, envoyImage string)
 					Labels:      envoyPodLabels(contour),
 				},
 				Spec: corev1.PodSpec{
-					// TODO anti-affinity
-					Affinity:       nil,
+					Affinity: &corev1.Affinity{
+						PodAntiAffinity: &corev1.PodAntiAffinity{
+							RequiredDuringSchedulingIgnoredDuringExecution: []corev1.PodAffinityTerm{
```

Comment:
Just wondering if we should switch this to `PreferredDuringSchedulingIgnoredDuringExecution` with a Weight of 100 to match what we do for the control plane Deployment (https://github.com/projectcontour/contour/blob/29201bbd13e5cf3477e35a8744f53bb13befb3b5/internal/provisioner/objects/deployment/deployment.go#L229-L243) -- it's a little more forgiving for single-node clusters, for example, it'll still schedule both replicas even if the anti-affinity rule can't be enforced.

I do see that the example envoy Deployment YAML uses "required", if we make this change then maybe we switch that as well for consistency.

I think I'd probably opt to go with `PreferredDuringSchedulingIgnoredDuringExecution` as the static default -- @sunjayBhatia @izturn any thoughts?

Reply:
IC, I could see in the test only 1 replica gets deployed, I thought it's just a middle state.

Reply:
```
matchExpressions:
                  - key: app
                    operator: In
                    values:
                      - envoy
```

The match expression can also be changed to `labelSelector` for consistency

Reply:
Using `PreferredDuringSchedulingIgnoredDuringExecution` seems good to me, though the upgrade from an existing Deployment using `RequiredDuringSchedulingIgnoredDuringExecution` might need a release note on the proper rollout process, since the new replicaset instances will be schedulable while the old ones are running. Tried updating a running Deployment just now and had to delete the old pods to get the new ones up on a one node cluster.

Reply:
MM with single node, even with `requiredDuringSchedulingIgnoredDuringExecution`, updating the container image will also run into the same issue. because old pod is not deleted, and it's still taking up the host port. new pod will get this error
```
  ----     ------            ----   ----               -------
  Warning  FailedScheduling  3m29s  default-scheduler  0/1 nodes are available: 1 node(s) didn't have free ports for the requested pod ports. preemption: 0/1 nodes are available: 1 No preemption victims found for incoming pod..
```


In order to fix this, either we ask customer have more nodes than envoy replicas, or we change the update strategy to Recreate, so it will make sure previous pod will be deleted first when updating the new antrea pod.
But I think it's unrelated to changing to `PreferredDuringSchedulingIgnoredDuringExecution `

Reply:
yeah when i tried it I removed the hostPorts from the Deployment containers since that is a non-starter, you can't start any new instance until the one taking the host port has been stopped

Reply:
I just added a log, not sure it's the best place to put it, please let me know if I need to move it somewhere else.